### PR TITLE
expression, executor: rewrite builtin function CONCAT_WS

### PR DESCRIPTION
**Pull request by winkyao**
rewrite builtin func using new expression evaluation architecture

### CODE REVIEW

**Comment by winkyao** [comment]
@XuHuaiyu @tiancaiamao PTAL

**Comment by hanfei1991** [comment]
We'd better to define a function type like `type InferTypeFunc func([]*FieldType) []*FieldType` and maintain a map containing `functionClassName -> InferTypeFunc` to reuse code.

**Comment by tiancaiamao** [comment]
LGTM

**Comment by XuHuaiyu** [comment]
@winkyao Please fix conflicts, and update the implementation since newBaseBuiltinFuncWithTp has been updated.

**Comment by winkyao** [comment]
@XuHuaiyu PTAL

File: expression/builtin_string.go [hunk 1]
```diff
@@ -232,44 +232,63 @@ type concatWSFunctionClass struct {
...
+		argTps = append(argTps, tpString)
+	}
+
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpString, argTps...)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	for i := range args {
+		bf.tp.Flag |= mysql.BinaryFlag
```
Comment (zz-jason) [inline expression/builtin_string.go]: only set binary flag when there is a binary parameter, and don't forget to set binary charset and collation once the result is a binary string

File: expression/builtin_string.go [hunk 2]
```diff
@@ -232,44 +232,63 @@ type concatWSFunctionClass struct {
...
+		bf.tp.Flen += args[i].GetType().Flen
```
Comment (zz-jason) [inline expression/builtin_string.go]: do over `mysql.MaxBlobWidth` check before addition

File: executor/executor_test.go [hunk 3]
```diff
@@ -811,6 +811,19 @@ func (s *testSuite) TestStringBuiltin(c *C) {
...
+	tk.MustExec("create table t(a int, b double, c datetime, d time, e char(20))")
+	tk.MustExec(`insert into t values(1, 1.1, "2017-01-01 12:01:01", "12:01:01", "abcdef")`)
+	result = tk.MustQuery("select concat_ws('|', a, b, c, d, e) from t")
+	result.Check(testkit.Rows("1|1.1|2017-01-01 12:01:01|12:01:01|abcdef"))
+	result = tk.MustQuery("select concat_ws(null, null)")
+	result.Check(testkit.Rows("<nil>"))
+	result = tk.MustQuery("select concat_ws(null, a, b) from t")
+	result.Check(testkit.Rows("<nil>"))
+	result = tk.MustQuery("select concat_ws(',', 'a', 'b')")
+	result.Check(testkit.Rows("a,b"))
```
Comment (zz-jason) [inline executor/executor_test.go]: add test for these situations:
1. separator is not NULL and some other parameters are NULL
2. with input parameter of binary type

File: expression/builtin_string.go [hunk 4]
```diff
@@ -232,44 +232,63 @@ type concatWSFunctionClass struct {
...
+			sep = val
 			continue
 		}
-		s = append(s, ss)
+		strs = append(strs, val)
 	}
 
-	d.SetString(strings.Join(s, sep))
-	return d, nil
+	return strings.Join(strs, sep), false, nil
```
Comment (zz-jason) [inline expression/builtin_string.go]: it's better to check weather the length of result string is larger than `Flen`

**Comment by winkyao** [comment]
@XuHuaiyu  @zz-jason  PTAL

File: expression/builtin_string_test.go [hunk 5]
```diff
@@ -191,27 +191,74 @@ func (s *testEvaluatorSuite) TestConcat(c *C) {
...
+		},
+	}
 
-	args = types.MakeDatums([]interface{}{errors.New("must error"), nil}...)
-	f, err = fc.getFunction(datumsToConstants(args), s.ctx)
-	c.Assert(err, IsNil)
-	v, err = f.eval(nil)
+	fcName := ast.ConcatWS
+	// ERROR 1582 (42000): Incorrect parameter count in the call to native function 'concat_ws'
+	f, err := newFunctionForTest(s.ctx, fcName, primitiveValsToConstants([]interface{}{nil})...)
```
Comment (XuHuaiyu) [inline expression/builtin_string_test.go]: parameter count must >= 2
which is defined in expression/builtin.go:
``` go
// funcs holds all registered builtin functions.
var funcs = map[string]functionClass{
...
	ast.ConcatWS:       &concatWSFunctionClass{baseFunctionClass{ast.ConcatWS, 2, -1}},
...
}
```

and be checked in 
``` go
func (c *concatWSFunctionClass) getFunction(){
errors.Trace(c.verifyArgs(args))
}
```

File: expression/builtin_string_test.go [hunk 6]
```diff
@@ -191,27 +191,74 @@ func (s *testEvaluatorSuite) TestConcat(c *C) {
...
 	c.Assert(err, NotNil)
```
Comment (XuHuaiyu) [inline expression/builtin_string_test.go]: what's this case for?

File: expression/builtin_string_test.go [hunk 7]
```diff
@@ -191,27 +191,74 @@ func (s *testEvaluatorSuite) TestConcat(c *C) {
...
+		} else {
+			c.Assert(err, IsNil)
+			if t.isNil {
+				c.Assert(val.Kind(), Equals, types.KindNull)
+			} else {
+				c.Assert(val.GetString(), Equals, t.expected)
+			}
+		}
+	}
 }
```
Comment (XuHuaiyu) [inline expression/builtin_string_test.go]: add test for isDetermistic

File: expression/builtin_string_test.go [hunk 8]
```diff
@@ -191,27 +191,74 @@ func (s *testEvaluatorSuite) TestConcat(c *C) {
...
-	args = types.MakeDatums([]interface{}{errors.New("must error"), nil}...)
-	f, err = fc.getFunction(datumsToConstants(args), s.ctx)
-	c.Assert(err, IsNil)
-	v, err = f.eval(nil)
+	fcName := ast.ConcatWS
+	// ERROR 1582 (42000): Incorrect parameter count in the call to native function 'concat_ws'
+	f, err := newFunctionForTest(s.ctx, fcName, primitiveValsToConstants([]interface{}{nil})...)
 	c.Assert(err, NotNil)
+
+	for _, t := range cases {
```
Comment (XuHuaiyu) [inline expression/builtin_string_test.go]: add test in plan/typeinfer_test.go/ TestInferType

File: expression/builtin_string_test.go (hunk 5 above)
Comment (winkyao) [inline expression/builtin_string_test.go]: yes, So I add a test case to test it will return error, when I pass param less than 2 param count

**Comment by winkyao** [comment]
@XuHuaiyu PTAL again...

File: expression/builtin_string.go [hunk 9]
```diff
@@ -200,8 +200,12 @@ func (c *concatFunctionClass) getFunction(args []Expression, ctx context.Context
 		return nil, errors.Trace(err)
 	}
 	for i := range args {
-		bf.tp.Flag |= mysql.BinaryFlag
-		bf.tp.Flen += args[i].GetType().Flen
+		argType := args[i].GetType()
+		if types.IsBinaryStr(argType) {
+			bf.tp.Flag |= mysql.BinaryFlag
```
Comment (zz-jason) [inline expression/builtin_string.go]: use `SetBinChsClnFlag(...)` to set flag/charset/collation to binary and break the execution of `for` loop

File: expression/builtin_string.go [hunk 10]
```diff
@@ -247,9 +251,17 @@ func (c *concatWSFunctionClass) getFunction(args []Expression, ctx context.Conte
 		if types.IsBinaryStr(argType) {
 			bf.tp.Flag |= mysql.BinaryFlag
```
Comment (zz-jason) [inline expression/builtin_string.go]: use `SetBinChsClnFlag(...)` to set flag/charset/collation to binary and break the execution of `for` loop

Comment (winkyao) [inline expression/builtin_string.go]: this loop is not just for binaryFlag set, it count the arg flen too.

File: expression/builtin_string.go (hunk 9 above)
Comment (winkyao) [inline expression/builtin_string.go]: this loop is not just for binaryFlag set, it count the arg flen too.

Comment (zz-jason) [inline expression/builtin_string.go]: i see

Comment (winkyao) [inline expression/builtin_string.go]: in the baseStringBuiltinFunc.getRetTp() will check the binary flag, and set charset and collation. @zz-jason 
```
func (b *baseStringBuiltinFunc) getRetTp() *types.FieldType {
    tp, flen := b.tp, b.tp.Flen
    if flen >= mysql.MaxBlobWidth {
        tp.Tp = mysql.TypeLongBlob
    } else if flen >= 65536 {
        tp.Tp = mysql.TypeMediumBlob
    }
    if mysql.HasBinaryFlag(tp.Flag) {
        tp.Charset, tp.Collate = charset.CharsetBin, charset.CollationBin
    } else {
        tp.Charset, tp.Collate = charset.CharsetUTF8, charset.CollationUTF8
    }
    return tp
}
```

**Comment by XuHuaiyu** [comment]
LGTM PTAL @zz-jason

**Pull request merged**